### DISCUSSION TITLE
EnsureUser: Check if user can be found

### DISCUSSION THREAD

**Discussion by bajce** [pnp/pnpcore#1062] [Q&A] [answered]
Is it possible to find out if the UPN of the user we want to 'ensure' is known to SharePoint instance?

I have following scenario:
Im programmatically creating user by Graph API via let say Application 1, UPN of the user is passed to Application 2 which should ensure user by PnPCoreSDK right after that but im getting the ServiceException with the message ''The specified user i:0#.f|membership|user-upn@tenant.com could not be found' and HttpResponse Code 400. 

Is there any way how to query for the user before ensure to check if the SharePoint knows this UPN? 
This would help me to build the mechanism to sleep before calling ensure. Checking the message of the exception is a little bit dangerous I think, as it can be changed or maybe it's not always in English and Response Code 400 is probably returned for all (or a lot of) bad requests. I can catch  all exceptions and retry but I think that is not the elegant solution. I could also always wait before calling `EnsureUser` but I also think that time how long it takes may differ.

**Comment by jansenbe** [comment, marked as answer]
@bajce : see https://github.com/pnp/pnpcore/issues/1030 and https://github.com/pnp/pnpcore/issues/1061.

**Reply by bajce** [reply]
@jansenbe happy to hear that  user validation will be supported, that's exactly what I need. As I workaround so far, im checking the `SeverErrorCode` of the `SharePointRestError`. Code `-2146232832L` is returned when user is not found. Im not sure that's the only case when  this code is returned when `EnsureUser` is called, but I think  its s till safer than checking the message content. Or do you see it different?

**Reply by jansenbe** [reply]
@bajce : EnsureUser is not entirely safe, if a user was never used for the site collection then it works fine and you'll see an error. However, when the user was previously added to the site collection, then removed from Azure AD calling EnsureUser would not return an error as it sees the user in the site collection user's table. That's why we're looking into adding a check that verifies at Azure AD level.

**Reply by bajce** [reply]
@jansenbe thank you for the clarification.
